### ISSUE TITLE
withFile memory usage

### ISSUE THREAD

**Issue by ludeus** [cakephp/cakephp#16701]
### Description

Downloading large files causes RAM issues.  
Indeed, the file is sent by consecutive "echo" in "emitBody" but if we are in a buffer level, the file is stored in RAM.

I think there is a bug in the call of "flush()", here is its call in the method "ResponseEmiter::emit()"

```php
        if (!headers_sent()) {
            $this->emitStatusLine($response);
            $this->emitHeaders($response);
        }
        $this->flush();

        $range = $this->parseContentRange($response->getHeaderLine('Content-Range'));
        if (is_array($range)) {
            $this->emitBodyRange($range, $response);
        } else {
            $this->emitBody($response);
        }
```

Looks like we want to flush the entire existing buffer before running "emitBody()"

But, the flush function is called with null param : 
```php
    protected function flush(?int $maxBufferLevel = null): void
    {
        if ($maxBufferLevel === null) {
            $maxBufferLevel = ob_get_level();
        }

        while (ob_get_level() > $maxBufferLevel) {
            ob_end_flush();
        }
    }
```
with a null "$maxBuffer Level", we don't go through the loop, so we might as well do nothing

### CakePHP Version

4.4.4

**Comment by markstory** [comment]
> with a null "$maxBuffer Level", we don't go through the loop, so we might as well do nothing

That's not entirely true. The application could have buffered output somewhere else and the current buffering level would be inferred.

In your scenario I'm assuming you have are starting an output buffer somewhere else or have automatic output buffering enabled?

**Comment by ndm2** [comment]
I think what he means is that this is essentially a no-op if no value is being provided explicitly, as `ob_get_level()` will never be greater than `ob_get_level()`.

If the buffer was meant to be completely flushed and disabled by default, then `$maxBufferLevel` would need to be set to `0`.

**Comment by markstory** [comment]
Fair enough, I don't think that is contributing to additional memory usage though.

**Comment by dereuromark** [comment]
Anything to be done here?
